shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -179,14 +181,15 @@ private void flushAll(ByteBuf out) {
             ArrayList<Session> sessions = new ArrayList<>(pendingWrites.values());
             sessions.forEach(session -> session.flush(out));
         }
+        tokenCounter.incrementAndGet();
```

Comment:
I don't understand why this token counter logic is being changed.

Reply:
in case of flush and close append; we flush the session of give writerid and segment. hence no need of increment the token counter for flush (one session ) ; its required for flushAll (all pending sessions)  triggered by blocktimeout or any wire command.

Reply:
What about the path on line 209? That does not increment the counter, so the block timeout will still occur and end up flushing some unrelated write that happens later unnecessarily.

Reply:
yes;  block timeout is required; theoretically , there may be pending data for other sessions.

Reply:
Ok, so there is a rather complex invariant we are trying to maintain here:
When appending there is only one session with a block open, and that sessions data goes out while others' accumulates. 
And if data has been waiting for more than a certain timeout, it gets flushed.

I am not exactly sure how best to do it. But it would be good if we could restructure this so that those are made more explicit. Perhaps checking against the session that owns the channel rather than a global token count? (With some sort of block id...)

Reply:
even I tried couple of experiments; but having separate timeouts each of writer ids sessions may load the client with timers. having single block timeout for multiple session would be less in load.

Reply:
Ok. Let's keeps it as a block. Below I think that should be:
```
            if (isChannelFree()) {
                if (session.isFree()) {
                    session.record(append);
                    startAppend(ctx, blockSizeSupplier, append, out);
                    continueAppend(data, out);
                    if (bytesLeftInBlock == 0) {
                        completeAppend(null, out);
                        tokenCounter.incrementAndGet();   // <--- added
                    }
                } else {
                  //... Is this branch reachable? 
                  //... Maybe this should throw an IllegalStateException?
                }
             }
```

Reply:
'''
 if (isChannelFree()) {
                if (session.isFree()) {
                    session.record(append);
                    startAppend(ctx, blockSizeSupplier, append, out);
                    continueAppend(data, out);
                    if (bytesLeftInBlock == 0) {
                        completeAppend(null, out);
                        tokenCounter.incrementAndGet();   // <--- added
 
here, while active session is writing on wire; other sessions accumulates in queue (pendingWrites).  The block timeout is required for pendingWrites

                    }
                } else {


                  //... Is this branch reachable? 
                  //... Maybe this should throw an IllegalStateException?

Yes, this branch is reachable and its valid case;  if the append completes and channel becomes free ; then if the write comes on a session in which there is already accumulated data exists , at this time , new data is append to accumulated data and flushed to wire.

                }
             }
'''

Reply:
Ok. lets add coverage for this case.

Reply:
@tkaitchuck 
Yes, I have added the test coverage now.. please let me know if you have any further comments.